src/auth/pluggable-auth-handler.ts

```diff
@@ -0,0 +1,208 @@
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import {ExecutableError} from './pluggable-auth-client';
+import {
+  ExecutableResponse,
+  ExecutableResponseError,
+  ExecutableResponseJson,
+  InvalidExpirationTimeFieldError,
+} from './executable-response';
+import * as childProcess from 'child_process';
+import * as fs from 'fs';
+
+/**
+ * Defines the options used for the PluggableAuthHandler class.
+ */
+export interface PluggableAuthHandlerOptions {
+  /**
+   * The command used to retrieve the third party token.
+   */
+  command: string;
+  /**
+   * The timeout in milliseconds for running executable,
+   * set to default if none provided.
+   */
+  timeoutMillis: number;
+  /**
+   * The path to file to check for cached executable response.
+   */
+  outputFile?: string;
+}
+
+/**
+ * A handler used to retrieve 3rd party token responses from user defined
+ * executables and cached file output for the PluggableAuthClient class.
+ */
+export class PluggableAuthHandler {
+  private readonly commandComponents: Array<string>;
+  private readonly timeoutMillis: number;
+  private readonly outputFile?: string;
+
+  /**
+   * Instantiates a PluggableAuthHandler instance using the provided
+   * PluggableAuthHandlerOptions object.
+   */
+  constructor(options: PluggableAuthHandlerOptions) {
+    if (!options.command) {
+      throw new Error('No command provided.');
+    }
+    this.commandComponents = PluggableAuthHandler.parseCommand(
+      options.command
+    ) as Array<string>;
+    this.timeoutMillis = options.timeoutMillis;
+    if (!this.timeoutMillis) {
+      throw new Error('No timeoutMillis provided.');
+    }
+    this.outputFile = options.outputFile;
+  }
+
+  /**
+   * Calls user provided executable to get a 3rd party subject token and
+   * returns the response.
+   * @param envMap a Map of additional Environment Variables required for
+   *   the executable.
+   * @return A promise that resolves with the executable response.
+   */
+  retrieveResponseFromExecutable(
```

Comment:
Let's do the GOOGLE_EXTERNAL_ACCOUNT_ALLOW_EXECUTABLES check here and not in the pluggable-auth-client since this is where executable handling happens

Reply:
That would change the behavior. If that check is moved, the output file will be checked and returned even if the env variable is not set. We don't want to return anything if the check fails right?

Reply:
I think I covered that in a different comment (having a method that does both and doing it there) but we discussed offline.